### PR TITLE
Chapter 9: Advanced Login

### PR DESCRIPTION
**Pull request by hazwanizafirah**
ready

### CODE REVIEW

**Comment by hazwanizafirah** [comment]
ready

File: app/models/user.rb [hunk 1]
```diff
@@ -9,4 +10,29 @@ class User < ApplicationRecord
...
     format: {with: VALID_EMAIL_REGEX}
 
+  class << self
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+
+    def digest string
+      cost = if ActiveModel::SecurePassword.min_cost
+        BCrypt::Engine.cost
```
Comment (anlvq-1058) [inline app/models/user.rb:20]:
```suggestion
      cost = if ActiveModel::SecurePassword.min_cost
               BCrypt::Engine::MIN_COST
            else
              BCrypt::Engine.cost
            end
```

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -4,15 +4,41 @@ def log_in user
...
 
   def current_user
-    @current_user ||= User.find_by id: session[:user_id]
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by id: user_id
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by id: user_id
+      if user&.authenticated?(cookies[:remember_token])
+        log_in user
+      @current_user = user
```
[the remark this answers is not in the document]
Comment (anlvq-1058) [reply app/helpers/sessions_helper.rb:13]: please fix this comment

**Review by anlvq-1058 — COMMENTED** [review]
cmt

**Comment by hazwanizafirah** [comment]
ready

**Comment by hazwanizafirah** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**